### PR TITLE
Record pleg pod relist interval and latency

### PR DESCRIPTION
**Pull request by yujuhong**
Relisting latency/interval affects how quick kubelet discovers changes. Record
the metrics in Prometheus to surface such information.

### CODE REVIEW

**Comment by yujuhong** [comment]
@k8s-bot unit test this

File: pkg/kubelet/pleg/generic.go [hunk 1]
```diff
@@ -101,6 +104,12 @@ func generateEvent(podID types.UID, cid string, oldState, newState kubecontainer
 // with the internal pods/containers, and generats events accordingly.
 func (g *GenericPLEG) relist() {
 	glog.V(5).Infof("GenericPLEG: Relisting")
+	timestamp := time.Now()
+	metrics.PLEGRelistInterval.Observe(metrics.SinceInMicroseconds(g.lastRelistTime))
```
Comment (Random-Liu) [inline pkg/kubelet/pleg/generic.go]: Just make sure, do we need to initialize `lastRelistTime` here?

Comment (yifan-gu) [inline pkg/kubelet/pleg/generic.go]: +1

File: pkg/kubelet/pleg/generic.go [hunk 2]
```diff
@@ -138,4 +147,7 @@ func (g *GenericPLEG) relist() {
 	for i := range events {
 		g.eventChannel <- events[i]
 	}
+
+	// Update the relist time.
+	g.lastRelistTime = timestamp
```
Comment (yifan-gu) [inline pkg/kubelet/pleg/generic.go]: Maybe move this up after `metrics.PLEGRelistInterval.Observe()` or in the defer clause ?

**Comment by yifan-gu** [comment]
Small nits, LGTM

File: pkg/kubelet/pleg/generic.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/pleg/generic.go]: Moving into defer clause maybe better, but moving up after `metrics.PLEGRelistInterval.Observe()` seems semantically different.

**Comment by Random-Liu** [comment]
LGTM with small nits.

File: pkg/kubelet/pleg/generic.go (hunk 2 above)
Comment (yifan-gu) [inline pkg/kubelet/pleg/generic.go]: @yujuhong Do you intend to count only successful relist?

File: pkg/kubelet/pleg/generic.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/pleg/generic.go]: Done. skipping the recording if it's not set.

File: pkg/kubelet/pleg/generic.go (hunk 2 above)
Comment (yujuhong) [inline pkg/kubelet/pleg/generic.go]: Done.

**Comment by yifan-gu** [comment]
@k8s-bot test this please

**Comment by yifan-gu** [comment]
@k8s-bot test this please

**Comment by yifan-gu** [comment]
@k8s-bot test this please

**Comment by wojtek-t** [comment]
"MetricsGrabber should grab all metrics from a Kubelet." this test is failing - can you please fix it?

**Comment by gmarek** [comment]
@yujuhong 
I added a tool for grabbing metrics and in the same time test which, I hope, will help us with managing metrics we gather. To fix this test you just need to add your metric names:
```
21:42:10           "kubelet_pleg_relist_latency_microseconds": {},
21:42:10           "kubelet_pleg_relist_latency_microseconds_sum": {},
21:42:10           "kubelet_pleg_relist_latency_microseconds_count": {},
21:42:10           "kubelet_pleg_relist_interval_microseconds": {},
21:42:10           "kubelet_pleg_relist_interval_microseconds_sum": {},
21:42:10           "kubelet_pleg_relist_interval_microseconds_count": {},
```
to `KnownKubeletMetrics` in `pkg/metrics/kubelet_metrics.go` together with all label names on which you're aggregating.

**Comment by Random-Liu** [comment]
@yujuhong A small fix is needed, :)

**Comment by yujuhong** [comment]
Added the metrics in `kubelet_metrics.go`. @gmarek, thanks for pointing out the fix.

@gmarek, would you mind clarifying the goal of metrics grabber? I couldn't find any explanation in the original PR. I imagine that it should ensure all metrics we want are present, but should not fail if there are extra/unknown metrics. However, it seems like it expects the exact list of metrics to be present. This forces anyone who touches the metrics to modify this centralized go file, causing unnecessary churn.

**Comment by gmarek** [comment]
We want to have a tool for looking for regression causes automatically. One of signals we want to use are metrics exposed by various components. Metrics grabber is just a single interface over all of them.

As a 'bonus' use I want to make it easier for owners to monitor all the metrics their components expose, by listing them in a single place.

**Pull request merged**